runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkStateInternals.java

```diff
@@ -215,6 +216,22 @@ public void clearGlobalState() {
     }
   }
 
+  public List<ByteBuffer> getGlobalWindowStateKeys() {
```

Comment:
nit
```suggestion
  @SuppressWarnings("unchecked")
  public List<ByteBuffer> getGlobalWindowStateKeys() {
```

Reply:
This will materialize all of the keys in-memory, which could be a problem with really large state (when using RocksDB).

Also since we're collecting keys into list, there will be duplicates if we have multiple states per key 🤔 (so we could expire a single window multiple times)

That being said, I don't know how to solve the "materialization" problem in combination with de-duplication :( (it would be fairly easy if we'd have only a single descriptor)

So for know we should at least replace list with set to address the possible duplicates. Also adding a failing test case for this would be great.

Reply:
Hmm second thought, we could probably simply pick a single state descriptor that we know is always present (some of the system ones, eg. watermark hold). Then we should be able to expire keys one by one.

Reply:
Good point on the deduplication. 

Are we guaranteed that the Flink runner will always have watermark hold stored for every key?

Reply:
I'm not 100% sure, as far as I remember the logic was done by the ReduceFnRunner + WatermarkHold. If things didn't change significantly, my best guess would be that is should work.

@je-ik any ideas?

Reply:
I would warn against picking random type of state and declare it invariant, that it must be present for all keys. Even if it works today, it might be subject to change and hard-to-debug errors. Although watermark hold looks like a pretty good candidate, I would pick it only as a last resort option.

The call `flinkStateBackend.applyToAllKeys` looks like it would deserve a better name, if it is applied to all "keys and states". :) The optimization could be done there, maybe the state backend knows it has sorted keys, and can therefore do the deduplication cheaper.

Until then I would suggest explicit deduplication, that should be O(n log n), which should be fine.

Reply:
@je-ik the concern wasn't the O() performance, rather the fact that we must materialize all keys in memory to deduplicate, which risks crashes.

Reply:
Ah, my bad.

In the case of memory complexity, I think there are few solutions, that would not be a sort of overengineering - without the proof there exists a need of such a complex solution. IMO, the cleanest would be to enhance the Flink state iteration API to provide appropriate guarantees (e.g. sorted keys), and deduplicate in-mem until such API is available.

Reply:
Oh you were talking about enhancing the current Flink API. I'm pretty sure we want to that, it could actually clash with some on-going efforts and tie our hands in further development of state backends

Reply:
Yes, I mean API, that would have the sorting (or unique keys, more likely) part of the contract.

Reply:
> Oh you were talking about enhancing the current Flink API. I'm pretty sure we don't want to do that, it could actually clash with some on-going efforts (changelog state backend) and tie our hands in further development of state backends

Why's that? Having an API like `iterateUniqueKeys`, seems to be able to have a default (in-mem) implementation, while it could be optimized for cases where the backend has sorted keys.

Reply:
I would say `org.apache.flink.runtime.state.KeyedStateBackend#getKeys` [1] should already provide everything you need. You just need to make sure you have a state that you can rely on.

If you think about it, you'd also need to make there is _some_ state in the approach you're proposing, so it's not that different.

[1] https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/runtime/state/KeyedStateBackend.html#getKeys-java.lang.String-N-

Reply:
I don't think there is any state, that would be guaranteed to be present for every window. That would require such state to be part of the Beam model and AFAIK there is no such state. Watermark hold might definitely be absent when there is nothing holding the watermark (no timer with output timestamp).
What comes close is window GC timer state, but even that can be optimized out for some cases (e.g. GlobalWindow). I think we should not rely on any internal detail, because that would be fragile.

Reply:
`iterateUniqueKeys` doesn't solve the issue then, because you can not rely on having AT LEAST SOME state for the key either.

For example, for some types of joins you might have cleaned up all state already before EOW 🤷‍♂️ (the key is no longer present in the underlying state backend)

Relying on internal detail that you have under control and that you can test for is IMO a better option that relying on how users write their DoFns..

Reply:
ok, it sounds like the simplest approach for now is to continue the current approach of deduplicating? (FYI this codepath here is specifically for the global window, so there will be no GC timer state)

Reply:
I don't know internal details of Flink. If there is no reliable way to get list of unique windows (which feels weird, but maybe it is possible), then we should simply resort to in-memory set. Either way, that is the cleanest for this PR, because any changes in Flink API would be long-term.

Reply:
Fixed deduplication - added comment to figure out a better way

Reply:
@reuvenlax It may not be enough. Take the bellow DoFn as an example. If we can not rely on having an internal state for each key that we've seen, wouldn't the example simply break?

```
new DoFn<KV<String, String>, Void> {

  void processFn(KV<String, String> element) {
    // write element to DB
  }

  @OnWindowExpiration
  void expire() {
    // delete element from DB... because you know, we can :)
  }
}
```

I have a feeling that we either:
- need to rely on state we know we always have
- we can not optimize EOGW gc timers in case we want OnExpiration callback and rely on those


Other "broken" example would be something along this lines...

```
new DoFn<KV<String, String>, Void> {

  BagState<String> buf;

  void processFn(KV<String, String> element) {
   if (buf.size() > 3) {
     buf.forEach(db::write);
     buf.clear();
   } else {
     buf.add(element);
   }
  }

  @OnWindowExpiration
  void expire() {
     // Flush remaining elements ... if we have any
     buf.forEach(db::write);
  }
}
```

Reply:
Maybe it's best to just set GC timers in this case (which is what Dataflow does) and do this in the GC timer. This will also alleviate the memory concerns.

Reply:
:-)
That's spot on. I think we need to cache seen windows (key-aligned in common key) in the state backend. Looks like we cannot reliably restore the set of ''active' windows after recovery from checkpoint.

Reply:
ok, I'm revert to the GC timer approach (was my first approach, but then I decided to "optimize" it).